substra/sdk/schemas.py

```diff
@@ -0,0 +1,267 @@
+# limitations under the License.
+
+import abc
+import contextlib
+import enum
+import typing
+from typing import Optional, List
+
+import pydantic
+
+from substra.sdk import utils
+
+# TODO create a sub-package schemas:
+# types
+# inputs
+# outputs
+
+
+_SERVER_NAMES = {
+    'dataset': 'data_manager',
+}
+
+
+class Type(enum.Enum):
+    Algo = 'algo'
+    AggregateAlgo = 'aggregate_algo'
+    CompositeAlgo = 'composite_algo'
+    DataSample = 'data_sample'
+    Dataset = 'dataset'
+    Model = 'model'
+    Objective = 'objective'
+    Testtuple = 'testtuple'
+    Traintuple = 'traintuple'
+    Aggregatetuple = 'aggregatetuple'
+    CompositeTraintuple = 'composite_traintuple'
+    ComputePlan = 'compute_plan'
+    Node = 'node'
+
+    def to_server(self):
+        """Returns the name used to identify the asset on the backend."""
+        name = self.value
+        return _SERVER_NAMES.get(name, name)
+
+    def __str__(self):
+        return self.name
+
+
+class _Spec(pydantic.BaseModel, abc.ABC):
+    """Asset creation specification base class."""
+
+    class Meta:
+        file_attributes = None
+
+    def is_many(self):
+        return False
+
+    @contextlib.contextmanager
+    def build_request_kwargs(self):
+        # TODO should be located in the backends/remote module
+        data = self.dict(exclude_unset=True)
+        if self.Meta.file_attributes:
+            with utils.extract_files(data, self.Meta.file_attributes) as (data, files):
+                yield (data, files)
+        else:
+            yield data, None
+
+
+class Permissions(pydantic.BaseModel):
+    public: bool
+    authorized_ids: typing.List[str]
+
+
+class PrivatePermissions(pydantic.BaseModel):
+    authorized_ids: typing.List[str]
+
+
+class DataSampleSpec(_Spec):
+    path: Optional[str]
+    paths: Optional[List[str]]
+    test_only: bool
+    data_manager_keys: typing.List[str]
+
+    type_: typing.ClassVar[Type] = Type.DataSample
+
+    def is_many(self):
+        return self.paths and len(self.paths) > 0
+
+    @pydantic.root_validator(pre=True)
+    def exclusive_paths(cls, values):
+        """Check that one and only one path(s) field is defined."""
+        if 'paths' in values and 'path' in values:
+            raise ValueError("'path' and 'paths' fields are exclusive.")
+        if 'paths' not in values and 'path' not in values:
+            raise ValueError("'path' or 'paths' field must be set.")
+        return values
+
+    @contextlib.contextmanager
+    def build_request_kwargs(self, local):
+        # redefine kwargs builder to handle the local paths
+        data = self.dict(exclude_unset=True)
+        if local:
+            with utils.extract_data_sample_files(data) as (data, files):
+                yield (data, files)
+        else:
+            yield data, None
+
+
+class DatasetSpec(_Spec):
+    name: str
+    data_opener: str
+    type: str
+    description: str
+    permissions: Permissions
+    objective_key: Optional[str]
```

Comment:
Since the beginning of this PR, metadata have been added to all assets (but datasamples), they should be added in the specs as well.

Reply:
Should it be added to 'models' as well ?
The type is Optional[dict] I suppose ?

Reply:
In specs, the type is `Optional[Dict[str,str]]` and in models it is `Dict[str,str]`: optional when submitted, but always a dict when returned by the backend (an empty dict in case of no value submitted).